File: Server/DealnetPortal.Api.Common/Helpers/LoanCalculator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Microsoft.VisualBasic;

namespace DealnetPortal.Api.Common.Helpers
{
    public static class LoanCalculator
    {
        public class Input
        {
            public double TaxRate { get; set; } 
            public int LoanTerm { get; set; }
            public int AmortizationTerm { get; set; }
            public double CustomerRate { get; set; }
            //Price of equipment(s) excl tax 
            public double PriceOfEquipment { get; set; }
            public double AdminFee { get; set; }
            public double DownPayment { get; set; }
            //can be as input data (for clarity program)
            public bool? IsClarity { get; set; }
            public bool IsOldClarityDeal { get; set; }
        }

        public class Output
        {
            public double Hst { get; set; }
            //Price of equipment (incl tax)
            public double PriceOfEquipmentWithHst { get; set; }
            public double TotalAmountFinanced { get; set; }
            //TODO: why double, not decimal?
            //TODO: because methods from Financial class outputs value in double
            public double TotalMonthlyPayment { get; set; }
            public double TotalAllMonthlyPayments { get; set; }
            public double ResidualBalance { get; set; }
            public double TotalObligation { get; set; }
            public double TotalBorowingCost { get; set; }

            public double LoanTotalCashPrice { get; set; }
        }

        public static Output Calculate(Input input)
        {            
            var output = new Output();
            output.Hst = input.TaxRate/100*input.PriceOfEquipment;
            output.PriceOfEquipmentWithHst = output.Hst + input.PriceOfEquipment;
            var customerRate = input.CustomerRate / 100 / 12;
            var mco = 0.0;
            if (input.IsClarity == true)
            {
                if (input.IsOldClarityDeal)
                {
                    output.TotalAmountFinanced = input.PriceOfEquipment /*+ input.AdminFee*/ - input.DownPayment;
                    //output.PriceOfEquipmentWithHst /*+ input.AdminFee*/ - input.DownPayment;
                    output.TotalMonthlyPayment = customerRate == 0 && input.AmortizationTerm == 0 ? 0 :
                        customerRate > 0 ? Math.Round(output.TotalAmountFinanced * Financial.Pmt(customerRate, input.AmortizationTerm, -1), 2)
                            : output.TotalAmountFinanced * Financial.Pmt(customerRate, input.AmortizationTerm, -1);
                    output.TotalAllMonthlyPayments = Math.Round(output.TotalMonthlyPayment, 2) * input.LoanTerm;
                    mco = output.TotalMonthlyPayment;
                }
                else
                {
                    const double clarityPaymentFactor = 0.010257;
                    output.TotalMonthlyPayment = output.PriceOfEquipmentWithHst;                             
                    output.PriceOfEquipmentWithHst = output.TotalMonthlyPayment / clarityPaymentFactor;
                    output.TotalAmountFinanced = output.PriceOfEquipmentWithHst + input.AdminFee - input.DownPayment;
                    mco = output.TotalAmountFinanced * clarityPaymentFactor;
                    output.TotalAllMonthlyPayments = Math.Round(mco, 2) * input.LoanTerm;                    
                }
            }
            else
            {
                output.TotalAmountFinanced = input.PriceOfEquipment /*+ input.AdminFee*/ - input.DownPayment;//output.PriceOfEquipmentWithHst /*+ input.AdminFee*/ - input.DownPayment;
                output.TotalMonthlyPayment = customerRate == 0 && input.AmortizationTerm == 0 ? 0 :
                    customerRate > 0 ? Math.Round(output.TotalAmountFinanced * Financial.Pmt(customerRate, input.AmortizationTerm, -1), 2)
                        : output.TotalAmountFinanced * Financial.Pmt(customerRate, input.AmortizationTerm, -1);
                output.TotalAllMonthlyPayments = Math.Round(output.TotalMonthlyPayment, 2) * input.LoanTerm;
                mco = output.TotalMonthlyPayment;
            }
            output.LoanTotalCashPrice = output.TotalAmountFinanced - input.AdminFee + input.DownPayment;
            if (input.LoanTerm != input.AmortizationTerm)
            {
                output.ResidualBalance = Math.Round(-Financial.PV(customerRate, input.AmortizationTerm - input.LoanTerm, mco) * (1 + customerRate),2);
            }
            output.TotalObligation = output.ResidualBalance + output.TotalAllMonthlyPayments + input.AdminFee;
            output.TotalBorowingCost = Math.Round(output.TotalObligation - output.TotalAmountFinanced - input.AdminFee,2);
            return output;
        }
    }
}

File: Client/DealnetPortal.Web/Scripts/utils/loan-calculator.js
function calculateLoanValues() {
    var hstLabel = $("#hst");
    var totalCashPriceLabel = $("#total-cash-price");
    var totalAmountFinancedLabel = $("#total-amount-financed");
    var loanTotalMonthlyPaymentLabel = $("#loan-total-monthly-payment");
    var loanTotalAllMonthlyPaymentsLabel = $("#loan-total-all-monthly-payments");
    var residualBalanceLabel = $("#residual-balance");
    var totalObligationLabel = $("#total-obligation");
    var totalBorrowingCostLabel = $("#total-borrowing-cost");
    hstLabel.text('-');
    totalCashPriceLabel.text('-');
    totalAmountFinancedLabel.text('-');
    loanTotalMonthlyPaymentLabel.text('-');
    loanTotalAllMonthlyPaymentsLabel.text('-');
    residualBalanceLabel.text('-');
    totalObligationLabel.text('-');
    totalBorrowingCostLabel.text('-');
    //
    var equipmentCashPrice = parseFloat($("#equipment-cash-price").text());
    if (isNaN(equipmentCashPrice) || equipmentCashPrice <= 0) {
        return;
    }
    var hst = 0;
    //Matt's request to map new contract
    //var hst = taxRate / 100 * equipmentCashPrice;
    hstLabel.text(formatNumber(hst));
    var totalCashPrice = equipmentCashPrice + hst;
    totalCashPriceLabel.text(formatNumber(totalCashPrice));
    var isFeePaidByCustomer = $('#fee-paid-by-customer').val().toLowerCase() == 'true';
    var adminFee = parseFloat($("#admin-fee").val());
    if (isNaN(adminFee) || adminFee < 0 || !isFeePaidByCustomer) {
        adminFee = 0;
    }
    var downPayment = parseFloat($("#down-payment").val());
    if (isNaN(downPayment) || downPayment < 0) {
        downPayment = 0;
    }
    var totalAmountFinanced = totalCashPrice - downPayment + adminFee;
    totalAmountFinancedLabel.text(formatNumber(totalAmountFinanced));

    var loanTerm = parseInt($("#loan-term").val());
    var amortizationTerm = parseInt($("#amortization-term").val());
    var customerRate = parseFloat($("#customer-rate").val());
    if (isNaN(loanTerm) || loanTerm <= 0 || isNaN(amortizationTerm) || amortizationTerm <= 0 || isNaN(customerRate) || customerRate < 0) {
        return;
    }
    var totalMonthlyPayment = customerRate > 0 ? (Math.round(totalAmountFinanced * pmt(customerRate / 100 / 12, amortizationTerm, -1, 0, 0) * 100) / 100) : (totalAmountFinanced * pmt(customerRate / 100 / 12, amortizationTerm, -1, 0, 0));
    isCalculationValid = totalMonthlyPayment > 0;
    loanTotalMonthlyPaymentLabel.text(formatNumber(totalMonthlyPayment));

    var totalAllMonthlyPayments = totalMonthlyPayment * loanTerm;
    loanTotalAllMonthlyPaymentsLabel.text(formatNumber(totalAllMonthlyPayments));

    var residualBalance = 0;
    if (loanTerm !== amortizationTerm) {
        residualBalance = -pv(customerRate / 100 / 12, amortizationTerm - loanTerm, totalMonthlyPayment, 0);
    }
    residualBalanceLabel.text(formatNumber(residualBalance));

    var totalObligation = totalAllMonthlyPayments + residualBalance;
    totalObligationLabel.text(formatNumber(totalObligation));

    var totalBorrowingCost = Math.abs(totalObligation - totalAmountFinanced + adminFee);
    totalBorrowingCostLabel.text(formatNumber(totalBorrowingCost));
}

function checkCalculationValidity(inputCashPrice, inputTaxRate) {
    var equipmentCashPrice = typeof inputCashPrice !== 'undefined' && inputCashPrice !== null ? inputCashPrice : parseFloat($("#equipment-cash-price").text());
    if (isNaN(equipmentCashPrice) || equipmentCashPrice <= 0) {
        return false;
    }
    var testTaxRate = typeof inputTaxRate !== 'undefined' && inputTaxRate !== null ? inputTaxRate : taxRate;
    var hst = testTaxRate / 100 * equipmentCashPrice;
    var totalCashPrice = equipmentCashPrice + hst;
    var adminFee = parseFloat($("#admin-fee").val());
    if (isNaN(adminFee) || adminFee < 0) {
        adminFee = 0;
    }
    var downPayment = parseFloat($("#down-payment").val());
    if (isNaN(downPayment) || downPayment < 0) {
        downPayment = 0;
    }
    var totalAmountFinanced = totalCashPrice - downPayment;
    var amortizationTerm = parseInt($("#amortization-term").val());
    var customerRate = parseFloat($("#customer-rate").val());
    if (isNaN(amortizationTerm) || amortizationTerm <= 0 || isNaN(customerRate) || customerRate < 0) {
        return false;
    }
    var totalMonthlyPayment = Math.round(totalAmountFinanced * pmt(customerRate / 100 / 12, amortizationTerm, -1, 0, 0) * 100) / 100;
    return totalMonthlyPayment > 0;
}

function calculateClarityTotalCashPrice() {
    var clarityCalcualtions = module.require('financial-functions.clarity');
    var hstLabel = $("#hst");
    var totalAmountFinancedLabel = $("#displayTotalAmount");
    var totalMonthlyPaymentsLabel = $("#displayAllMonthly");
    var displayDealerCost = $("#displayDealerCost");
    var balanceOwning = $("#displayBalanceOwning");
    var totalObligationLabel = $("#displayTotalObl");
    var borrowerCost = $("#displayCostOfBorrow");
    var yourCost = $("#displayYourCost");
    var loantTermDisplay = $("#displayLoanTerm");
    var amortTermDisplay = $("#displayAmortTem");
    var displayCustRate = $('#displayCustRate');
    var displayMonthlyCostNoTax = $('#totalMonthlyCostNoTax');
    var displayTotalHst = $('#total-hst');
    var displayTotalMonthlyCostTax = $('#totalMonthlyCostTax');
    var displaytotalMonthlyCostTaxDP = $('#totalMonthlyCostTaxDP')

    hstLabel.text('-');
    totalAmountFinancedLabel.text('-');
    totalMonthlyPaymentsLabel.text('-');
    totalObligationLabel.text('-');
    borrowerCost.text('-');
    yourCost.text('-');
    balanceOwning.text('-');
    displayMonthlyCostNoTax.text('-');
    displayTotalHst.text('-');
    displayTotalMonthlyCostTax.text('-');
    displaytotalMonthlyCostTaxDP.text('-');

    var equipmentCashPrice = parseFloat($("#equipment-cash-price").text());
    var packageCashPrice = parseFloat($("#package-cash-price").text());
    var hst = taxRate;
    var loanTerm = parseInt($("#loan-term").val());
    var amortizationTerm = parseInt($("#amortization-term").val());
    var customerRate = parseFloat($("#customer-rate").val());
    var adminFee = parseFloat($("#admin-fee").val());
    if (isNaN(adminFee) || adminFee < 0) {
        adminFee = 0;
    }
    var downPayment = parseFloat($("#down-payment").val());
    if (isNaN(downPayment) || downPayment < 0) {
        downPayment = 0;
    }

    var data = {
        tax: hst,
        equipmentSum: equipmentCashPrice,
        packagesSum: packageCashPrice,
        LoanTerm: loanTerm,
        AmortizationTerm: amortizationTerm,
        CustomerRate: customerRate,
        AdminFee: adminFee,
        downPayment: downPayment,
        DealerCost: 0
    }

    hstLabel.text(formatNumber(clarityCalcualtions.tax(data)));
    totalAmountFinancedLabel.text(formatNumber(clarityCalcualtions.totalAmountFinanced(data)));
    totalMonthlyPaymentsLabel.text(formatNumber(clarityCalcualtions.totalMonthlyPayments(data)));
    totalObligationLabel.text(formatNumber(clarityCalcualtions.totalObligation(data)));
    borrowerCost.text(formatNumber(clarityCalcualtions.totalBorrowingCost(data)));
    balanceOwning.text(formatNumber(clarityCalcualtions.residualBalance(data)));
    yourCost.text(formatNumber(clarityCalcualtions.yourCost(data)));
    displayCustRate.text(customerRate + ' %');
    loantTermDisplay.text(loanTerm + ' months');
    amortTermDisplay.text(amortizationTerm + ' months');
    amortTermDisplay.text(amortizationTerm + ' months');
    displayDealerCost.text(0 + ' %');
    displayMonthlyCostNoTax.text(formatNumber(clarityCalcualtions.totalMCONoTax(data)));
    displayTotalHst.text(formatNumber(clarityCalcualtions.tax(data)));
    displaytotalMonthlyCostTaxDP.text(formatNumber(clarityCalcualtions.totalMonthlyPaymentsLessDownPayment(data)));

    displayTotalMonthlyCostTax.text(formatNumber(clarityCalcualtions.totalMCOLessDownPaymentNoTax(data)));

}